amaranth_soc/gpio.py

```diff
@@ -0,0 +1,333 @@
+
+from . import csr
+
+
+__all__ = ["PinMode", "PinSignature", "Peripheral"]
+
+
+class PinMode(enum.Enum, shape=unsigned(2)):
+    """GPIO pin mode.
+
+    The 2-bit values of this enumeration can be written to a :class:`Peripheral.Mode` field to
+    configure the pins of a :class:`Peripheral`.
+    """
+
+    #: `Input-only` mode.
+    #:
+    #: The pin output is disabled but remains connected to its :class:`Peripheral.Output` field.
+    #: Its :attr:`Peripheral.alt_mode` bit is wired to 0.
+    INPUT_ONLY = 0b00
+
+    #: `Push-pull` mode.
+    #:
+    #: The pin output is enabled and connected to its :class:`Peripheral.Output` field. Its
+    #: :attr:`Peripheral.alt_mode` bit is wired to 0.
+    PUSH_PULL = 0b01
+
+    #: `Open-drain` mode.
+    #:
+    #: The pin output is enabled when the value of its :class:`Peripheral.Output` field is 0, and
+    #: is itself wired to 0. Its :attr:`Peripheral.alt_mode` bit is wired to 0.
+    OPEN_DRAIN = 0b10
+
+    #: `Alternate` mode.
+    #:
+    #: The pin output is disabled but remains connected to its :class:`Peripheral.Output` field.
+    #: Its :attr:`Peripheral.alt_mode` bit is wired to 1.
+    ALTERNATE = 0b11
+
+
+class PinSignature(wiring.Signature):
+    """GPIO pin signature.
+
+    Interface attributes
+    --------------------
+    i : :class:`Signal`
+        Input.
+    o : :class:`Signal`
+        Output.
+    oe : :class:`Signal`
+        Output enable.
+    """
+    def __init__(self):
+        super().__init__({
+            "i":  In(unsigned(1)),
+            "o":  Out(unsigned(1)),
+            "oe": Out(unsigned(1)),
+        })
+
+
+class Peripheral(wiring.Component):
+    class Mode(csr.Register, access="rw"):
+        """Mode register.
+
+        This :class:`csr.Register` contains an array of ``pin_count`` read/write fields. Each field
+        is 2-bit wide and its possible values are defined by the :class:`PinMode` enumeration.
+
+        If ``pin_count`` is 4, then the register has the following fields:
+
+        .. bitfield::
+            :bits: 8
+
+                [
+                    { "name": "pin[0]", "bits": 2, "attr": "RW" },
+                    { "name": "pin[1]", "bits": 2, "attr": "RW" },
+                    { "name": "pin[2]", "bits": 2, "attr": "RW" },
+                    { "name": "pin[3]", "bits": 2, "attr": "RW" },
+                ]
+
+        Parameters
+        ----------
+        pin_count : :class:`int`
+            Number of GPIO pins.
+        """
+        def __init__(self, pin_count):
+            super().__init__({
+                "pin": [csr.Field(csr.action.RW, PinMode) for _ in range(pin_count)],
+            })
+
+    class Input(csr.Register, access="r"):
+        """Input register.
+
+        This :class:`csr.Register` contains an array of ``pin_count`` read-only fields. Each field
+        is 1-bit wide and driven by the input of its associated pin in the :attr:`Peripheral.pins`
+        array.
```

Comment:
Should probably mention the synchronizer.

Reply:
Done.